bionic/task_state.py

```diff
@@ -335,13 +329,37 @@ def initialize(self, bootstrap, flow_instance_uuid):
             for task_key in self.task_keys
         ]
 
-        should_persist = (
-            single_unique_element(
-                entity_def.can_persist
-                for entity_def in self.entity_defs_by_dnode.values()
-            )
-            and not self.output_would_be_missing()
+        # In theory different entities for a single task could have different cache
+        # settings, but I'm not sure it can happen in practice (given the way
+        # grouped entities are created). At any rate, once we have tuple
+        # descriptors, each task state will only be responsible for a single entity
+        # and this won't be an issue.
+        should_memoize, should_persist = single_unique_element(
+            (entity_def.should_memoize, entity_def.can_persist)
+            for entity_def in self.entity_defs_by_dnode.values()
         )
+
+        if should_memoize is None:
+            if bootstrap is None:
+                should_memoize = True
+            else:
+                should_memoize = bootstrap.default_memoization
+        if self.provider.attrs.changes_per_run and not should_memoize:
+            descriptors = [
+                task_key.dnode.to_descriptor() for task_key in self.task_keys
+            ]
+            message = f"""
+            Descriptors {descriptors!r} aren't set to be memoized but are
+            decorated with @changes_per_run. This isn't allowed as
+            @changes_per_run computes once in a flow instance and memoizes
+            the value. The results will be memoized for it to run correctly.
+            To avoid this warning, enable memoization for the descriptor by
+            applying `@memoize(True)` to the corresponding function."""
+            warnings.warn(oneline(message))
+            should_memoize = True
+        self.should_memoize = should_memoize
+
+        should_persist = should_persist and not self.output_would_be_missing()
```

Comment:
Interesting point. I think about it the same way for mutating class variables, but for local variables like these, I feel like it should be okay since all the updates to the local variables are co-located. You will see that I'm setting the class variables only once. 

Is it confusing because there is a different code section (separated by newline) for initially setting them and later updating them? In which case, I can see why we would rename the first variables to something like `opt_should_memoize` and `can_persist`.

Reply:
Yes, I think it's because there's another code section (not related to `should_persist`) in between. It also seems better when it's sort of a safety override, like in the `should_persist` case. In this case, it's more that we just have complicated logic, which IMO should be all in one place.